### PR TITLE
Add `ModeSolver.solve_group_index` (#819)

### PR DESCRIPTION
**Pull request by lucas-flexcompute**
Calculation was added to `ModeSolver`, because it is not always required, so it doesn't make much sense to pay the price of the calculation if we're just interested in the effective index. I don't know if there's a better way to to it.

### CODE REVIEW

File: tidy3d/plugins/mode/mode_solver.py [hunk 1]
```diff
@@ -116,6 +116,58 @@ def solve(self) -> ModeSolverData:
...
+        frac_step : float
+            Fractional frequency step used for numerical differentiation.
+
+        Returns
+        -------
+        :class:`ModeIndexDataArray`
+            Group indices for all waveguide wavelengths and modes indices.
+        """
+        if frac_step >= 1:
+            raise ValidationError("Argument `frac_step` must be less than 1.")
```
Comment (momchil-flex) [inline tidy3d/plugins/mode/mode_solver.py:137]: We use single quotes, i.e. `'frac_step'` formatting in error/warning messages.

File: tidy3d/plugins/mode/mode_solver.py [hunk 2]
```diff
@@ -116,6 +116,58 @@ def solve(self) -> ModeSolverData:
...
+        # of mode tracking.
+        freqs = np.outer(self.freqs, (1 - frac_step, 1, 1 + frac_step)).flatten()
+        end = freqs.size
+
+        mode_solver = self.copy(update={"freqs": freqs})
+        n_backward = mode_solver.data.n_eff.isel(f=slice(0, end, 3)).values
+        n_center = mode_solver.data.n_eff.isel(f=slice(1, end, 3)).values
+        n_forward = mode_solver.data.n_eff.isel(f=slice(2, end, 3)).values
+
+        n_group = n_center + (n_forward - n_backward) / (2 * frac_step)
```
Comment (momchil-flex) [inline tidy3d/plugins/mode/mode_solver.py:161]: I guess if we want to save 1/3 of the solves we could also do

```
n_center = (n_forward + n_backward) / 2
n_group = n_center + (n_forward - n_backward) / (2 * frac_step)
```

I would think this should not have any noticeable effect on precision, what do you think?

**Review by momchil-flex — COMMENTED** [review]
I think this is generally OK, but an alternative could also be to just have a `compute_group_index=False` kwarg in `ModeSolver.solve`?

Also, could you add a test?

**Comment by lucas-flexcompute** [comment]
> I think this is generally OK, but an alternative could also be to just have a `compute_group_index=False` kwarg in `ModeSolver.solve`?

In that case, do you mean to add a field like `n_group` to `ModeSolverDataset`, such that when `compute_group_index == True` it gets calculated, but otherwise it is set to `None`?

File: tidy3d/plugins/mode/mode_solver.py (hunk 2 above)
Comment (lucas-flexcompute) [reply tidy3d/plugins/mode/mode_solver.py:161]: Honestly I'm not a big fan of this… It's hard to say how much of an impact it has (each waveguide and wavelength region would be impacted differently) but there's definitely some impact. I think a 3-point derivative is not that expensive and gives reasonable results (I actually though about using a 5-point derivative, but given the errors in the n_eff, I thought it might be too expensive for no real gain).
Now, I know I'm kind of conservative with these things and tend to err on the side of caution, so my opinion might not be what most users want…

Comment (momchil-flex) [reply tidy3d/plugins/mode/mode_solver.py:161]: Fair enough, also it may be moot depending on what we decide about whether to have it as part of the regular `solve`, or as a standalone `solve_group_index`.

**Comment by momchil-flex** [comment]
> > I think this is generally OK, but an alternative could also be to just have a `compute_group_index=False` kwarg in `ModeSolver.solve`?
> 
> In that case, do you mean to add a field like `n_group` to `ModeSolverDataset`, such that when `compute_group_index == True` it gets calculated, but otherwise it is set to `None`?

What do you think @tylerflex ? Have a separate call to compute the group index, or optionally compute it during the regular `solve`, but then have the `n_group` part of `ModeSolverDataset` be `None` by default, unless group index is requested? This doesn't sound too bad to me?

One other point: if we do it this way, we could also add the `compute_group_index` kwarg to the `ModeSolverMonitor`, and have a way to get the group index with subpixel averaging.

**Comment by momchil-flex** [comment]
In some sense, I feel like we should scrap `ModeSolverMonitor` altogether, and just have `ModeMonitor` with flags for `store_amps`, `store_effective_index`, `store_fields`, `store_group_index`... Where the first two are on by default and the last two are off by default. I remember we considered something like that a while ago but opted for the ModeSolverMonitor I believe to better separate things. But I kinda like the flag method all things considered. For example, it also won't necessitate two separate mode solves for a ModeMonitor and a ModeSolverMonitor at the same place, if that's what the user wants. Am I missing some downside @tylerflex ?

**Comment by lucas-flexcompute** [comment]
I can see one possibly unexpected issue. The user won't be able to change his mind after solving because we cache the data:
```python
data = mode_solver.solve()
assert data.n_group is None
new_data = mode_solver.solver(compute_group_index=True)
assert new_data.n_group is None
```

**Comment by tylerflex** [comment]
As discussed, my vote is to keep the group index calculation as a separate method. I see potential pitfalls with changing the resulting data based on kwargs. It seems much more straightforward to me to add an optional method and then not have to guess / check whether the group index data is present. This was similar to the reasoning behind changing the monitors to always store the same data, instead of toggling based on the `store` keyword if you remember that @momchil-flex.

**Comment by lucas-flexcompute** [comment]
If that's the final decision, I've already included a test for the calculation and fixed the warning message, so this should be ready to merge.

File: tidy3d/plugins/mode/mode_solver.py [hunk 3]
```diff
@@ -71,6 +82,30 @@ def freqs_not_empty(cls, val):
...
+            return GROUP_INDEX_STEP
+        return val
+
+    def _post_init_validators(self):
+        if self.compute_group_index >= 1:
+            raise ValidationError("Argument 'compute_group_index' must be less than 1.")
+
+        if self.compute_group_index > 0:
+            if self.mode_spec.track_freq is None:
+                log.warning(
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:98]: I think it would improve usability if we could tell the user specifically what field to set in these warnings. For example, by adding the sentence `Consider setting the 'ModeSpec.track_freq' field to __.` Same for the precision. `Consider setting the 'ModeSpec.precision' field to 'double'`.

File: tidy3d/plugins/mode/mode_solver.py [hunk 4]
```diff
@@ -116,6 +151,7 @@ def solve(self) -> ModeSolverData:
         """
         return self.data
 
+    # pylint: disable=too-many-locals
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:154]: I think it could improve readability to break this method up. or at least the group index portion of it. is it possible? Then I took a look at the full `data_raw` and noticed that it's already very long and hard to follow. @momchil-flex maybe this is pretty low priority but the mode solver code probably could be cleaned up a bit in the future. @lucas-flexcompute I'll leave it to you to decide if it's worth it, but I do think throwing a separate method to compute group index could really improve readability

```py
def data_raw(self) -> ModeSolverData:
    if self.compute_group_index > 0:
        return self._compute_group_index()
    ...
```

File: tidy3d/plugins/mode/mode_solver.py [hunk 5]
```diff
@@ -71,6 +82,30 @@ def freqs_not_empty(cls, val):
...
 
+    @pydantic.validator("compute_group_index")
+    def assign_default_on_true(cls, val):
+        """Assing the default fractional frequency step value if not provided."""
+        if val is True:
+            return GROUP_INDEX_STEP
+        return val
+
+    def _post_init_validators(self):
+        if self.compute_group_index >= 1:
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:93]: might be better to throw this in the `assign_default_on_true` validator and it could be caught earlier. For example
```py
@pydantic.validator("compute_group_index")
def assign_default_on_true(cls, val):
    """Assing the default fractional frequency step value if not provided."""
    if val is True:
        val = GROUP_INDEX_STEP
    if val >= 1:
        raise ValidationError("Argument 'compute_group_index' must be less than 1.")
    return val
```

File: tidy3d/plugins/mode/mode_solver.py [hunk 6]
```diff
@@ -71,6 +82,30 @@ def freqs_not_empty(cls, val):
...
+            raise ValidationError("Argument 'compute_group_index' must be less than 1.")
+
+        if self.compute_group_index > 0:
+            if self.mode_spec.track_freq is None:
+                log.warning(
+                    "Group index calculation without mode tracking can lead to incorrect results "
+                    "around mode crossings."
+                )
+
+            if self.mode_spec.precision != "double":
```
Comment (tylerflex) [inline tidy3d/plugins/mode/mode_solver.py:103]: might be better to do this as a root validator

```py
@pd.root_validator()
def check_precision(self, values):
     if values["compute_group_index"] and values["precision"] != "double":
         log.warning(...)
```

**Review by tylerflex — CHANGES_REQUESTED** [review]
just a few changes, otherwise looks good. thanks!

File: tidy3d/plugins/mode/mode_solver.py (hunk 4 above)
Comment (lucas-flexcompute) [reply tidy3d/plugins/mode/mode_solver.py:154]: Sure thing, I've separated the group index calculation as you suggested.

**Review by tylerflex — APPROVED** [review]
